### PR TITLE
Add new `unflatten` implementation

### PR DESCRIPTION
**Pull request by jachris**
This PR changes the behavior of `unflatten` and should hopefully be more stable. Take a look at 2d1aaa0685d953680f493fa3fc6e424e893c38a8 to compare the behavior.

### CODE REVIEW

File: libtenzir/include/tenzir/table_slice.hpp [hunk 1]
```diff
@@ -402,6 +402,9 @@ auto resolve_operand(const table_slice& slice, const operand& op)
 auto unflatten(const table_slice& slice,
                std::string_view nested_field_separator) -> table_slice;
 
+/// Just like `unflatten`, but... better? worse?
+auto unflatten2(const table_slice& slice, std::string_view sep) -> table_slice;
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/table_slice.hpp:406]: As much as I laughed about the joke, I think we should actually describe the difference here and maybe even choose to get rid of the old implementation. I think the new one better.

File: libtenzir/src/table_slice.cpp [hunk 2]
```diff
@@ -1302,4 +1301,241 @@ auto unflatten(const table_slice& slice,
...
+  }
+  auto array = check(to_record_batch(slice)->ToStructArray());
+  auto result = unflatten(array, sep);
+  auto cast = std::dynamic_pointer_cast<arrow::StructArray>(std::move(result));
+  TENZIR_ASSERT(cast);
+  auto schema = type{slice.schema().name(), type::from_arrow(*cast->type())};
+  // TODO: Is this what we want?
+  auto batch = arrow::RecordBatch::Make(schema.to_arrow_schema(),
+                                        cast->length(), cast->fields());
+  // TODO: Offset, import_time, etc.
```
Comment (dominiklohmann) [inline libtenzir/src/table_slice.cpp:1537]: Especially the import time is rather important for the compaction use case, given that that strongly relies on it.

I think the offset we could actually remove, it should be dead code by now.

File: libtenzir/src/table_slice.cpp [hunk 3]
```diff
@@ -1302,4 +1301,241 @@ auto unflatten(const table_slice& slice,
...
+  TENZIR_ASSERT(record);
+  return realize(std::move(*record));
+}
+
+auto unflatten(std::shared_ptr<arrow::Array> array, std::string_view sep)
+  -> std::shared_ptr<arrow::Array>;
+
+auto unflatten(const arrow::ListArray& array, std::string_view sep)
+  -> std::shared_ptr<arrow::ListArray> {
+  // TODO: array.offset?
```
Comment (dominiklohmann) [inline libtenzir/src/table_slice.cpp:1506]: Shouldn't that be handled by `array.values()`, or does that not take care of the offset?

File: libtenzir/src/table_slice.cpp [hunk 4]
```diff
@@ -1302,4 +1301,241 @@ auto unflatten(const table_slice& slice,
...
+auto realize(unflatten_entry&& entry) -> std::shared_ptr<arrow::Array> {
+  TENZIR_WARN("pre realize");
+  return entry.match<std::shared_ptr<arrow::Array>>(
+    [](std::shared_ptr<arrow::Array>& array) {
+      TENZIR_ASSERT(array);
+      TENZIR_WARN("realizing {}", array->type()->ToString());
+      return std::move(array);
+    },
+    [](unflatten_record& record) {
+      TENZIR_WARN("realizing record");
```
Comment (dominiklohmann) [inline libtenzir/src/table_slice.cpp:1473]: A few leftover debug prints like this, but I'm also reviewing while this is still in draft. Just leaving this one comment here as a reminder.

File: libtenzir/src/table_slice.cpp [hunk 5]
```diff
@@ -1302,4 +1301,241 @@ auto unflatten(const table_slice& slice,
...
+  auto ty = array.struct_type();
+  auto names = ty->fields() | std::views::transform(&arrow::Field::name);
+  // We need to flatten the null bitmap here because it can happen that the
+  // fields are saved to a record that is made non-null by another entry.
+  auto fields = check(array.Flatten());
+  if (not std::holds_alternative<unflatten_record>(root)) {
+    root.emplace<unflatten_record>(array.length(), array.null_bitmap());
+  }
+  for (auto [name, data] : detail::zip_equal(names, fields)) {
+    // TODO: Empty name? Double separators?
```
Comment (dominiklohmann) [inline libtenzir/src/table_slice.cpp:1425]: The double separators should result in an empty field name, right? I think the old implementation did not handle that correctly.

**Review by dominiklohmann — COMMENTED** [review]
Just a first look.

File: libtenzir/src/table_slice.cpp (hunk 3 above)
Comment (jachris) [reply libtenzir/src/table_slice.cpp:1506]: Apparently not:

```cpp
/// \brief Return array object containing the list's values
///
/// Note that this buffer does not account for any slice offset or length.
const std::shared_ptr<Array>& values() const { return values_; }
```

File: libtenzir/src/table_slice.cpp (hunk 5 above)
Comment (jachris) [reply libtenzir/src/table_slice.cpp:1115]: Seems like `std::views::split` does the right thing for double separators, but not for empty name. The empty name should still be the empty name after splitting, but it returns nothing (seems like a bad choice to me). Added a workaround for that.

**Review by dominiklohmann — APPROVED** [review]
This works nicely. Just three small comments.

**Pull request merged**